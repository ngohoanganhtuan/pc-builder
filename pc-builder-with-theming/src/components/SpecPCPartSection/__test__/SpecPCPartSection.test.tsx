import { BrowserRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

// Components
import { SpecPCPartSection } from '@components';

// Enums
import { BRAND_NAMES, PC_PART_TYPES } from '@enums';

describe('Test [SpecPCPartSection] component', () => {
  const handleRemovePCPart = jest.fn();
  const handleIncreaseQuantity = jest.fn();
  const handleDecreaseQuantity = jest.fn();
  const handleOpenModal = jest.fn();

  test('Component [SpecPCPartSection] should render correctly when it has PC part data', () => {
    const component = renderer
      .create(
        <BrowserRouter>
          <SpecPCPartSection
            pcPartTypeName={PC_PART_TYPES.CPU}
            pcPartData={{
              id: 0,
              name: 'Intel i3',
              brandName: BRAND_NAMES.INTEL,
              price: 1200000,
              thumbnail: 'PC Part Thumbnail',
              type: PC_PART_TYPES.CPU,
            }}
            quantity={500}
            handleRemovePCPart={handleRemovePCPart}
            handleIncreaseQuantity={handleIncreaseQuantity}
            handleDecreaseQuantity={handleDecreaseQuantity}
            handleOpenModal={handleOpenModal}
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Component [SpecPCPartSection] should render correctly when it does not have pc part data', () => {
    const component = renderer
      .create(
        <BrowserRouter>
          <SpecPCPartSection
            pcPartTypeName={PC_PART_TYPES.CPU}
            quantity={500}
            handleRemovePCPart={handleRemovePCPart}
            handleIncreaseQuantity={handleIncreaseQuantity}
            handleDecreaseQuantity={handleDecreaseQuantity}
            handleOpenModal={handleOpenModal}
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Component [SpecPCPartSection] calls function remove when user click remove button', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <SpecPCPartSection
          pcPartTypeName={PC_PART_TYPES.CPU}
          pcPartData={{
            id: 0,
            name: 'Intel i3',
            brandName: BRAND_NAMES.INTEL,
            price: 1200000,
            thumbnail: 'PC Part Thumbnail',
            type: PC_PART_TYPES.CPU,
          }}
          quantity={500}
          handleRemovePCPart={handleRemovePCPart}
          handleIncreaseQuantity={handleIncreaseQuantity}
          handleDecreaseQuantity={handleDecreaseQuantity}
          handleOpenModal={handleOpenModal}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('pc-part-remove-button'));
    expect(handleRemovePCPart).toHaveBeenCalled();
  });

  test('Component [SpecPCPartSection] calls function increase when user click increase button', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <SpecPCPartSection
          pcPartTypeName={PC_PART_TYPES.CPU}
          pcPartData={{
            id: 0,
            name: 'Intel i3',
            brandName: BRAND_NAMES.INTEL,
            price: 1200000,
            thumbnail: 'PC Part Thumbnail',
            type: PC_PART_TYPES.CPU,
          }}
          quantity={500}
          handleRemovePCPart={handleRemovePCPart}
          handleIncreaseQuantity={handleIncreaseQuantity}
          handleDecreaseQuantity={handleDecreaseQuantity}
          handleOpenModal={handleOpenModal}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('increase-quantity-button'));
    expect(handleIncreaseQuantity).toHaveBeenCalled();
  });

  test('Component [SpecPCPartSection] calls function decrease when user click decrease button', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <SpecPCPartSection
          pcPartTypeName={PC_PART_TYPES.CPU}
          pcPartData={{
            id: 0,
            name: 'Intel i3',
            brandName: BRAND_NAMES.INTEL,
            price: 1200000,
            thumbnail: 'PC Part Thumbnail',
            type: PC_PART_TYPES.CPU,
          }}
          quantity={500}
          handleRemovePCPart={handleRemovePCPart}
          handleIncreaseQuantity={handleIncreaseQuantity}
          handleDecreaseQuantity={handleDecreaseQuantity}
          handleOpenModal={handleOpenModal}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('decrease-quantity-button'));
    expect(handleDecreaseQuantity).toHaveBeenCalled();
  });

  test('Component [SpecPCPartSection] calls function open modal when user click select button', () => {
    const handleOpenModal = jest.fn();
    const { getByRole } = render(
      <BrowserRouter>
        <SpecPCPartSection
          pcPartTypeName={PC_PART_TYPES.CPU}
          quantity={500}
          handleRemovePCPart={handleRemovePCPart}
          handleIncreaseQuantity={handleIncreaseQuantity}
          handleDecreaseQuantity={handleDecreaseQuantity}
          handleOpenModal={handleOpenModal}
        />
      </BrowserRouter>
    );
    const button = getByRole('button');
    fireEvent.click(getByRole('button'));
    expect(button.textContent).toEqual('Select');
    expect(handleOpenModal).toHaveBeenCalled();
  });
});
