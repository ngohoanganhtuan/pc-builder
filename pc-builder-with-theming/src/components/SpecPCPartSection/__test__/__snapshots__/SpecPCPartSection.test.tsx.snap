// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test [SpecPCPartSection] component Component [SpecPCPartSection] should render correctly when it does not have pc part data 1`] = `
<div
  className="d-flex flex-row justify-content-between"
>
  <h2
    className="pc-part-section-name fs-5"
  >
    Processor (CPU)
  </h2>
  <button
    className="btn btn-primary btn-sm border-0 shadow"
    onClick={[Function]}
    type="button"
  >
    Select
  </button>
</div>
`;

exports[`Test [SpecPCPartSection] component Component [SpecPCPartSection] should render correctly when it has PC part data 1`] = `
<div
  className=""
>
  <h2
    className="pc-part-section-name fs-5"
  >
    Processor (CPU)
  </h2>
  <div
    className="shadow d-grid pc-part"
  >
    <div
      className="p-1"
    >
      <img
        className="border rounded-circle pc-part-thumbnail"
        data-testid="pc-part-thumbnail"
        height="100%"
        src="PC Part Thumbnail"
        width="100%"
      />
    </div>
    <div
      className="p-3 d-flex flex-column align-items-start justify-content-between"
    >
      <a
        className="name card-title fs-6 text-dark fw-bold"
        href="/detail/0"
        onClick={[Function]}
      >
        Intel i3
      </a>
      <div
        className="quantity-input d-flex input-group-sm shadow rounded-pill overflow-hidden quantity-input-sm"
      >
        <button
          className="btn bg-white shadow-none"
          data-testid="decrease-quantity-button"
          onClick={[Function]}
        >
          <i
            className="bi bi-dash"
          />
        </button>
        <div
          className="number bg-white d-flex align-items-center w-100 justify-content-center"
        >
          500
        </div>
        <button
          className="btn bg-white shadow-none"
          data-testid="increase-quantity-button"
          onClick={[Function]}
        >
          <i
            className="bi bi-plus"
          />
        </button>
      </div>
    </div>
    <div
      className="p-3 d-flex flex-column align-items-end justify-content-between"
    >
      <p
        className="fs-6 price"
      >
        1.200.000 ₫
      </p>
      <button
        className="btn btn-sm btn-danger shadow"
        data-testid="pc-part-remove-button"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
